### PR TITLE
http: add on-upstream-connected access log flush

### PR DESCRIPTION
**Pull request by ohadvano**
Commit Message: Add on-upstream-connected access log flush for HTTP Connection Manager.
Additional Description: HTTP version follow up for TCP Proxy feature: https://github.com/envoyproxy/envoy/pull/26094
Risk Level: Low
Testing: Unit tests
Docs Changes: Access logs
Release Notes: None
Platform Specific Features: None

Resolves https://github.com/envoyproxy/envoy/issues/26136

### CODE REVIEW

File: docs/root/intro/arch_overview/observability/access_logging.rst
Comment (paul-r-gall) [inline docs/root/intro/arch_overview/observability/access_logging.rst:29]: Add note that this will fire multiple times in the case of retries.

Comment (ohadvano) [reply docs/root/intro/arch_overview/observability/access_logging.rst:29]: The callback will be called upon successfully establishing L4 connection with the upstream, it's not considering L7 retries. Have I missed something?

Comment (ohadvano) [reply docs/root/intro/arch_overview/observability/access_logging.rst:29]: Do you mean that even though L4 connection was successful, L7 retries might cause a new L4 connection to be made? I see. Do you have a suggestion for how to phrase this?
Does the following make sense: "In case HTTP retry caused a new connection to be opened, multiple access logs will be recorded, per the number of new connections that were successfully established."

By the way, what will happen for end of stream access log in that case? Will there be only one or multiple?

Comment (paul-r-gall) [reply docs/root/intro/arch_overview/observability/access_logging.rst:29]: It actually may be worse than that, because even is a retry picks the exact same connection, it'll still call onPoolReady (i.e. the logic is: "let me make a new upstream request --> let me pick an upstream connection --> let me call onPoolReady when that's done"). 

End of Stream access logging is controlled in conn_manager_impl, and will fire exactly once per-stream.

Comment (paul-r-gall) [reply docs/root/intro/arch_overview/observability/access_logging.rst:29]: you might be better off implementing this on upstream_logs_: https://github.com/envoyproxy/envoy/blob/32a1f3586e5a7c84ff96ada4c7ecaee12dfdd745/source/common/router/upstream_request.cc#L209-L212

File: docs/root/intro/arch_overview/observability/access_logging.rst
Comment (alyssawilk) [inline docs/root/intro/arch_overview/observability/access_logging.rst:28]: established TCP connection -> established a connection
(for HTTP/3 it's not a TCP connection)

File: docs/root/intro/arch_overview/observability/access_logging.rst
Comment (alyssawilk) [reply docs/root/intro/arch_overview/observability/access_logging.rst:29]: if the concern is where to plumb the config, could we address that by allowing HCM config to turn up upstream access logging?
I agree with Matt and paul that having an upstream event trigger a downstream flush makes less sense to me than having the upstream event trigger an upstream log flush.

**Review by alyssawilk — COMMENTED** [review]
Thanks for tackling this and bearing with the design discussion :-)

File: docs/root/intro/arch_overview/observability/access_logging.rst
Comment (ohadvano) [reply docs/root/intro/arch_overview/observability/access_logging.rst:29]: As I see it, using upstream access logs would not solve the retry confusion? It's just another method for having access logs. I'm still not sure about upstream access logs. To be honest I wasn't even aware it exists until Paul brought it up, and even if the config propagates from HCM to the router, the user still needs to define the access log schema all over again in two instances (HCM + router). IMO this is even more confusing, because the user would enable that config and expect access logs to be written but the format and data would turn out different than as defined for HCM.

Comment (mattklein123) [reply docs/root/intro/arch_overview/observability/access_logging.rst:29]: I think for your use case you should be using upstream access logs. If we need a similar set of options for the upstream logs around periodic, emit at beginning, etc. I would add config for that directly.

Comment (ohadvano) [reply docs/root/intro/arch_overview/observability/access_logging.rst:29]: @mattklein123, seems like upstream access logs are also being recorded when the stream is ended, how does it solve this issue for logging on stream start?

Comment (mattklein123) [reply docs/root/intro/arch_overview/observability/access_logging.rst:29]: It does not. 

> If we need a similar set of options for the upstream logs around periodic, emit at beginning, etc. I would add config for that directly.

Comment (mattklein123) [reply docs/root/intro/arch_overview/observability/access_logging.rst:29]: I'm not sure quite what you mean. Downstream/upstream logs have separate configs. If they are duplicated I think that is just an implication of your deployment?

Comment (ohadvano) [reply docs/root/intro/arch_overview/observability/access_logging.rst:29]: I meant that I think it would be better to utilize the downstream access log config. Feels to me that they are more "mainstream" and usually used more frequent than upstream logs. However, I do understand your point. Do you think it would make more sense to add this for upstream access logs? If yes, then for each upstream connection, including retries, or should it make more sense to log on response headers that has success status code?

Comment (ohadvano) [reply docs/root/intro/arch_overview/observability/access_logging.rst:29]: Ok, Before I do this, what's practically the difference between downstream and upstream access logs? Wouldn't the outcome be the same in terms of how the logs would look and when they are recorded?

Comment (mattklein123) [reply docs/root/intro/arch_overview/observability/access_logging.rst:29]: Yes

Comment (ohadvano) [reply docs/root/intro/arch_overview/observability/access_logging.rst:29]: I understand now. I think adding this feature (and also the periodic logging) to upstream access log is important. Nevertheless, in my particular case (or generally speaking), I wouldn't want to have multiple access logs per stream (even considering retries). Only one at the end of stream and one in the beginning of a stream, so I would have to use the downstream access log. However now I do acknowledge that current implementation does not make sense.

In that case, would it make sense to modify current implementation to either:
- Record access log on stream start (before even going to router filter), to indicate that a request has been received, even though that at this stage, some info would be unknown, such as upstream host, as it wasn't selected yet.
- Record access log when response headers have returned and indicated success status code

Both cases should be accessible to HCM (downstream part), so it does not require modifying upstream implementation to access data on the downstream.

Comment (mattklein123) [reply docs/root/intro/arch_overview/observability/access_logging.rst:29]: Yup sure that would be fine also.

File: test/common/http/conn_manager_impl_test.cc
Comment (paul-r-gall) [inline test/common/http/conn_manager_impl_test.cc:2427]: nit: I don't think that the Times(2) is necessary -- the multiple WillOnce is sufficient.

**Review by paul-r-gall — CHANGES_REQUESTED** [review]

File: test/common/http/conn_manager_impl_test.cc
Comment (paul-r-gall) [reply test/common/http/conn_manager_impl_test.cc:2432]: right; stream state doesn't exist yet. would it be ok to merge this one after https://github.com/envoyproxy/envoy/pull/26256, so that we can add the tests for this?

File: test/common/http/conn_manager_impl_test.cc
Comment (ohadvano) [reply test/common/http/conn_manager_impl_test.cc]: done

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto
Comment (ohadvano) [reply api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto]: done

File: test/common/http/conn_manager_impl_test.cc
Comment (ohadvano) [reply test/common/http/conn_manager_impl_test.cc:2431]: Sure

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto
Comment (mattklein123) [reply api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:559]: +1 I agree, let's do this now.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Generally LGTM modulo the API comment.

/wait

**Comment by ohadvano** [comment]
@paul-r-gall, @mattklein123 I intend to add similar features to upstream access logs and therefore would it be ok if we don't block this PR on this change and instead I'll open an Issue to batch these API changes under a single PR for tcp_proxy, HCM and router's access logs at once?

**Review by mattklein123 — APPROVED** [review]

**Pull request merged**
